dgc/Sign1CoseMessage.cs

```diff
@@ -0,0 +1,279 @@
+        public string KID { get; private set; }
+        public CBORObject ProtectedMap { get; private set; }
+
+        public static Sign1CoseMessage DecodeFromBytes(byte[] coseBytes)
+        { 
+            var cborMsg = CBORObject.DecodeFromBytes(coseBytes);
+
+            if (cborMsg.Type != CBORType.Array) throw new InvalidDataException("Message is not a COSE security message.");
+            if (cborMsg.MostOuterTag.ToInt32Checked() != Sign1Tag) throw new InvalidDataException("Message is not a COSE security message.");
+            if (cborMsg.Count != 4) throw new InvalidDataException("Invalid Sign1 structure");
+
+            var protectedBytes = cborMsg[0].GetByteString();
+            var protectedMap = CBORObject.DecodeFromBytes(protectedBytes);
+            
+            var unprotectedMap = cborMsg[1];
+
+            var coseMsg = new Sign1CoseMessage();
+            coseMsg.Content = cborMsg[2].GetByteString();
+            coseMsg.Signature = cborMsg[3].GetByteString();
+            coseMsg.ProtectedMap = protectedMap;
+
+            var algKey = protectedMap[HeaderKey_Alg];
+            if (algKey.AsInt32() == Alg_ES256.AsInt32())
+            {
+                coseMsg.RegisteredAlgorithm = HCertSupportedAlgorithm.ES256;
+            }
+            else if (algKey.AsInt32() == Alg_PS256.AsInt32())
+            {
+                coseMsg.RegisteredAlgorithm = HCertSupportedAlgorithm.PS256;
+            }
+            else
+            {
+                throw new NotSupportedException("Algorithm not supported");
+            }
+
+            if (unprotectedMap[HeaderKey_KID] != null)
+            {
+                var kidBytes = unprotectedMap[HeaderKey_KID].GetByteString();
+                coseMsg.KID = Encoding.UTF8.GetString(kidBytes);
+            }
+            else
+            {
+                var kidBytes = protectedMap[HeaderKey_KID].GetByteString();
+                coseMsg.KID = Encoding.UTF8.GetString(kidBytes);
+            }
+            return coseMsg;
+        }
+
+        public byte[] EncodeToBytes()
+        {
+            var coseObj = CBORObject.NewArray();
+
+            var protectedMap = CBORObject.NewMap();
+            if (RegisteredAlgorithm == HCertSupportedAlgorithm.ES256)
+            {
+                protectedMap[HeaderKey_Alg] = Alg_ES256;
+            }
+            else if (RegisteredAlgorithm == HCertSupportedAlgorithm.PS256)
+            {
+                protectedMap[HeaderKey_Alg] = Alg_PS256;
+            }
+
+            coseObj.Add(protectedMap.EncodeToBytes());
+
+            var unProtectedMap = CBORObject.NewMap();
+            var kidBytes = Encoding.UTF8.GetBytes(KID);
+            unProtectedMap[HeaderKey_KID] = CBORObject.FromObject(kidBytes);
+            coseObj.Add(unProtectedMap);
+
+            coseObj.Add(CBORObject.FromObjectAndTag(Content, 2));
+            coseObj.Add(CBORObject.FromObjectAndTag(Signature, 3));
```

Comment:
2, 3, -> magic numbers, similar to above -> const or enum or equiv telling me what they are